feat(docs): page on migrating remark -> mdx (#23832)
gatsbyjs/gatsby: docs/docs/mdx/migrate-remark-to-mdx.md

---
title: Migrating Remark to MDX
---

For people who already have an existing blog using gatsby-transformer-remark, but want to use MDX, you can swap out the remark transformer plugin with gatsby-plugin-mdx and touch little code otherwise.

## Prerequisites

- An existing site that builds pages using `gatsby-transformer-remark` ([gatsby-starter-blog](https://github.com/gatsbyjs/gatsby-starter-blog) will be used as a reference in this guide).

## Adding in gatsby-plugin-mdx

Add gatsby-plugin-mdx (and its peer dependencies) to your package.json and remove gatsby-transformer-remark.

```shell
npm install @mdx-js/mdx @mdx-js/react gatsby-plugin-mdx
npm remove gatsby-transformer-remark
```

## Replacing gatsby-transformer-remark with gatsby-plugin-mdx

> In your gatsby-config.js file, replace gatsby-transformer-remark with gatsby-plugin-mdx. All sub-plugins of gatsby-transformer-remark can still work with gatsby-plugin-mdx by updating the plugins option to gatsbyRemarkPlugins.

Reviewer: I worry about saying all here. There is at least one I'm aware of that has been known to clash. Maybe saying that you can use remark plugins with mdx without the all piece?

> In your gatsby-config.js file, replace gatsby-transformer-remark with gatsby-plugin-mdx. All sub-plugins of gatsby-transformer-remark can still work with gatsby-plugin-mdx by updating the plugins option to gatsbyRemarkPlugins.

Reviewer: Sounds good to me.

```diff:title=gatsby-config.js
{
- resolve: `gatsby-transformer-remark`
+ resolve: `gatsby-plugin-mdx`
  options: {
-   plugins: [
+   gatsbyRemarkPlugins: [
```

## Update gatsby-node.js

In the `createPages` API call, when you query for `allMarkdownRemark`, replace it with `allMdx`.

```diff:title=gatsby-node.js
const result = await graphql(
  `
    {
-     allMarkdownRemark(
+     allMdx(
        sort: { fields: [frontmatter___date], order: DESC }
        limit: 1000
      ) {
```

Also, update `onCreateNode` which creates the blog post slugs to watch for the node type of `Mdx` instead of `MarkdownRemark`.

```diff:title=gatsby-node.js
exports.onCreateNode = ({ node, actions, getNode }) => {
  const { createNodeField } = actions

-  if (node.internal.type === `MarkdownRemark`) {
+  if (node.internal.type === `Mdx`) {
```

## Update File Extension

Where your markdown files live, change the file extension to `.mdx`.

> Also, you can tell `gatsby-plugin-mdx` to accept both `md` and `mdx` files by adding the `extensions` option in your gatsby-config entry.

Reviewer: I might move this up to the section changing gatsby-config stuff and make the extension recommendation the default. Then note that they also have the option of changing the file extensions directly.

Author: I moved this section up as a subsection of "Replacing gatsby-transformer-remark with gatsby-plugin-mdx" and cleaned it up a little based on @marcysutton's suggestions.

```js:title=gatsby-config.js
  {
    resolve: `gatsby-plugin-mdx`,
    options: {
      extensions: [`.md`, `.mdx`], // highlight-line
    },
  },
```

Now with this addition, gatsby-plugin-mdx will see files that end with both `.mdx` or `.md`.

## Update usage in pages

> Similar to gatsby-node, whenever you use `allMarkdownRemark` in a GraphQL query, change it to `allMdx`.

Reviewer 2: ```suggestion
Similar to `gatsby-node.js`, wherever you use `allMarkdownRemark` in a GraphQL query, change it to `allMdx`.
```

> Then in your blogpost template, to render the MDX, pull in the `MDXRenderer` react component from `gatsby-plugin-mdx`.

Reviewer 2: ```suggestion
Then in your blogpost template, to render the MDX, pull in the `MDXRenderer` React component from `gatsby-plugin-mdx`.
```

```js:title=src/templates/blog-post.js
import { MDXRenderer } from "gatsby-plugin-mdx"
```

> And in the graphql query, change the `html` field in `mdx` to `body`.

Reviewer 2: ```suggestion
And in the GraphQL query, change the `html` field in `mdx` to `body`.
```

```graphql:title=src/templates/blog-post.js
mdx(fields: { slug: { eq: $slug } }) {
  id
  excerpt(pruneLength: 160)
  body // highlight-line
  frontmatter {
    ...
  }
}
```

And finally swap out the component with `dangerouslySetInnerHTML` to a `MDXRenderer` component:

```diff:title=src/templates/blog-post.js
const post = data.mdx

// ...

-<section dangerouslySetInnerHTML={{ __html: post.html }} />
+<MDXRenderer>{post.body}</MDXRenderer>
```

> ## Update markdown files that includes HTML

Reviewer 2: ```suggestion
## Update Markdown files that include HTML code
```

> As MDX uses [JSX](/docs/glossary/jsx/) instead of HTML, examine anywhere you put in HTML and make sure that it is valid JSX.

Reviewer 2: ```suggestion
As MDX uses [JSX](/docs/glossary/jsx/) instead of HTML, examine anywhere you put in HTML and make sure that it is valid JSX. There might be some reserved words that need to be modified, or attributes to convert to camelCase.
```

For instance, any HTML component with the `class` attribute needs to be changed to `className`.

```diff
-<span class="highlight">Hello World</span>
+<span className="highlight">Hello World</span>
```

> ## Additional Resources:

Reviewer 2: ```suggestion
## Additional resources
```

## Update File Extension

Where your markdown files live, change the file extension to `.mdx`.

Also, you can tell `gatsby-plugin-mdx` to accept both `md` and `mdx` files by adding the `extensions` option in your gatsby-config entry.

```js
  {
    resolve: `gatsby-plugin-mdx`,
    options: {
      extensions: [`.md`, `.mdx`], // highlight-line
    },
  },
```

## Update usage in pages

Similar to gatsby-node, whenever you use `allMarkdownRemark` in a GraphQL query, change it to `allMdx`.

Then in your blogpost template, to render the MDX, pull in the `MDXRenderer` react component from `gatsby-plugin-mdx`.

```js
import { MDXRenderer } from "gatsby-plugin-mdx"
```

And in the graphql query, change the `html` field in `mdx` to `body`.

```graphql
mdx(fields: { slug: { eq: $slug } }) {
  id
  excerpt(pruneLength: 160)
  body
  frontmatter {
    ...
  }
}
```

And finally swap out the component with `dangerouslySetInnerHTML` to a `MDXRenderer` component:

```diff
const post = data.mdx

// ...

-<section dangerouslySetInnerHTML={{ __html: post.html }} />
+<MDXRenderer>{post.body}</MDXRenderer>
```

> 

Author: We decided it would be good to add in this section if you had any HTML in your markdown how to migrate it over as HTML in MDX is actually JSX.

Author: Done

## What's next?

> Go check out the [Importing and Using Components in MDX guide](/docs/mdx/importing-and-using-components) to find out how you can insert React components in your MDX files.

Author: Look into seeing if there are more resources that would be good to look at after this guide

Author: Also added another link on adding remark plugins
